{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "79cb28c6-3ddc-4020-a602-b5d93787c482",
   "metadata": {},
   "source": [
    "## General idea: Find out whether the genetics influence the microbiome by comparing the samples within the monozygotic and dizygotic pairs and find the significance of the beta diversity. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e6d3f765-d08f-4cad-9cfe-2e9ee718035e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing all required packages & notebook extensions at the start of the notebook\n",
    "import os\n",
    "import pandas as pd\n",
    "import qiime2 as q2\n",
    "from qiime2 import Visualization\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "from operator import itemgetter\n",
    "import matplotlib.patches as mpatches\n",
    "from scipy.stats import shapiro\n",
    "from os.path import exists\n",
    "\n",
    "or_dir = '../data' #original data (demux sequences, metadata)\n",
    "data_dir = 'data' #data from polybox (ASV, taxonomy analysis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fe9287ec-7166-41ad-a517-beaf4ae2910f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-11-27 10:36:30 URL:https://polybox.ethz.ch/index.php/s/pNA39R0rl2xMMj9/download [903525744] -> \"data.zip\" [1]\n"
     ]
    }
   ],
   "source": [
    "! wget -nv -O data.zip 'https://polybox.ethz.ch/index.php/s/pNA39R0rl2xMMj9/download'\n",
    "! unzip -q data.zip #-d $data_dir\n",
    "! mv data data2\n",
    "! mv data2/taxonomy/data .\n",
    "! cp data2/ASV/data/* data\n",
    "! cp data2/Phylogeny/data/* data\n",
    "! rm -r data2\n",
    "! rm data.zip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93648300-2b1e-4ad0-9902-8378a77bd88e",
   "metadata": {},
   "source": [
    "### 1. Separate metadata table for mono- and dizygotic twins and generate a table for each twin individually. (Or maybe for each pair?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d86f1c5f-b2ab-42ab-bcb0-30fc59a1a2b1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([42.1, 27.2, 28.1, 28.2, 39.2,  8.1,  8.2, 29.1, 40.1, 40.2, 35.1,\n",
       "       35.2, 47.1, 47.2,  4.1,  4.2, 29.2,  3.1, 30.2, 36.1, 36.2,  6.1,\n",
       "        6.2, 30.1, 33.1, 33.2, 43.2, 44.1, 44.2, 45.1, 45.2,  5.1, 37.1,\n",
       "       37.2, 39.1, 46.1,  3.2, 43.1, 42.2, 46.2,  5.2, 27.1, 48.2, 48.1,\n",
       "       32.1, 32.2, 12.2, 13.2, 14.1, 14.2, 10.1, 10.2, 12.1, 13.1, 15.1,\n",
       "       15.2, 16.1, 25.1, 25.2, 26.2, 11.1,  2.1,  2.2, 20.1, 20.2, 21.1,\n",
       "       21.2, 23.1, 23.2, 19.2, 16.2, 17.1, 17.2, 18.1, 18.2, 19.1, 24.2,\n",
       "       11.2, 24.1, 26.1])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pd.read_csv(or_dir + '/metadata.tsv', sep = '\\t')\n",
    "host_numbers = metadata['host_id'].unique()\n",
    "host_numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bb7de49d-3cce-4463-99d0-553c6aa93fb2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "all_hosts = dict()\n",
    "for host in host_numbers: #loop through all unique host ids\n",
    "    #print(host)\n",
    "    new_name = 'df_host_'+str(host)\n",
    "    #new_name = new_name.replace('.', '_')\n",
    "    all_hosts[host] = metadata[metadata['host_id']==host]\n",
    "    locals()[new_name] = metadata[metadata['host_id']==host]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7030fe8-b271-48ff-8fc1-1a6054966124",
   "metadata": {},
   "source": [
    "### 2. Problem: some samples contain NaN values, but the host has been weaned before. We need to keep those values and assign the status of weaned and lose all others that do not contain any information. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "a0dbe9b0-40e9-40e1-be4b-4b2e37b1773b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Library Layout</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geo_location_name</th>\n",
       "      <th>geo_latitude</th>\n",
       "      <th>geo_longitude</th>\n",
       "      <th>host_id</th>\n",
       "      <th>age_days</th>\n",
       "      <th>weight_kg</th>\n",
       "      <th>...</th>\n",
       "      <th>birth_length_cm</th>\n",
       "      <th>sex</th>\n",
       "      <th>delivery_mode</th>\n",
       "      <th>zygosity</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>delivery_preterm</th>\n",
       "      <th>diet_milk</th>\n",
       "      <th>diet_weaning</th>\n",
       "      <th>age_months</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1600</th>\n",
       "      <td>ERR1311612</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-06-09 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>36.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1589</th>\n",
       "      <td>ERR1311616</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-07-08 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>65.0</td>\n",
       "      <td>4.763</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1258</th>\n",
       "      <td>ERR1310030</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-11-10 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>190.0</td>\n",
       "      <td>6.804</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1259</th>\n",
       "      <td>ERR1310031</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-12-03 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>213.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904</th>\n",
       "      <td>ERR1310681</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-01-14 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>256.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>905</th>\n",
       "      <td>ERR1310682</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-02-17 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>290.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>906</th>\n",
       "      <td>ERR1310683</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-03-16 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>316.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1617</th>\n",
       "      <td>ERR1311611</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-05-05 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>366.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1587</th>\n",
       "      <td>ERR1311614</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-07-07 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>428.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1586</th>\n",
       "      <td>ERR1311613</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-08-05 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>457.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1588</th>\n",
       "      <td>ERR1311615</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-10-18 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>532.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1590</th>\n",
       "      <td>ERR1311617</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-11-15 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>560.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>907</th>\n",
       "      <td>ERR1310684</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-01-18 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>23.1</td>\n",
       "      <td>624.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id Library Layout      Instrument      collection_date  \\\n",
       "1600  ERR1311612         PAIRED  Illumina MiSeq  2010-06-09 00:00:00   \n",
       "1589  ERR1311616         PAIRED  Illumina MiSeq  2010-07-08 00:00:00   \n",
       "1258  ERR1310030         PAIRED  Illumina MiSeq  2010-11-10 00:00:00   \n",
       "1259  ERR1310031         PAIRED  Illumina MiSeq  2010-12-03 00:00:00   \n",
       "904   ERR1310681         PAIRED  Illumina MiSeq  2011-01-14 00:00:00   \n",
       "905   ERR1310682         PAIRED  Illumina MiSeq  2011-02-17 00:00:00   \n",
       "906   ERR1310683         PAIRED  Illumina MiSeq  2011-03-16 00:00:00   \n",
       "1617  ERR1311611         PAIRED  Illumina MiSeq  2011-05-05 00:00:00   \n",
       "1587  ERR1311614         PAIRED  Illumina MiSeq  2011-07-07 00:00:00   \n",
       "1586  ERR1311613         PAIRED  Illumina MiSeq  2011-08-05 00:00:00   \n",
       "1588  ERR1311615         PAIRED  Illumina MiSeq  2011-10-18 00:00:00   \n",
       "1590  ERR1311617         PAIRED  Illumina MiSeq  2011-11-15 00:00:00   \n",
       "907   ERR1310684         PAIRED  Illumina MiSeq  2012-01-18 00:00:00   \n",
       "\n",
       "             geo_location_name  geo_latitude  geo_longitude  host_id  \\\n",
       "1600  USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "1589  USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "1258  USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "1259  USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "904   USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "905   USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "906   USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "1617  USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "1587  USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "1586  USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "1588  USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "1590  USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "907   USA, Missouri, St. Louis      38.63699     -90.263794     23.1   \n",
       "\n",
       "      age_days  weight_kg  ...  birth_length_cm   sex  delivery_mode  \\\n",
       "1600      36.0        NaN  ...             47.0  male        Vaginal   \n",
       "1589      65.0      4.763  ...             47.0  male        Vaginal   \n",
       "1258     190.0      6.804  ...             47.0  male        Vaginal   \n",
       "1259     213.0        NaN  ...             47.0  male        Vaginal   \n",
       "904      256.0        NaN  ...             47.0  male        Vaginal   \n",
       "905      290.0        NaN  ...             47.0  male        Vaginal   \n",
       "906      316.0        NaN  ...             47.0  male        Vaginal   \n",
       "1617     366.0        NaN  ...             47.0  male        Vaginal   \n",
       "1587     428.0        NaN  ...             47.0  male        Vaginal   \n",
       "1586     457.0        NaN  ...             47.0  male        Vaginal   \n",
       "1588     532.0        NaN  ...             47.0  male        Vaginal   \n",
       "1590     560.0        NaN  ...             47.0  male        Vaginal   \n",
       "907      624.0        NaN  ...             47.0  male        Vaginal   \n",
       "\n",
       "       zygosity       race     ethnicity  delivery_preterm diet_milk  \\\n",
       "1600  Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "1589  Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "1258  Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "1259  Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "904   Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "905   Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "906   Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "1617  Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "1587  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1586  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1588  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1590  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "907   Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "\n",
       "     diet_weaning age_months  \n",
       "1600        False        1.0  \n",
       "1589        False        2.0  \n",
       "1258         True        6.0  \n",
       "1259         True        7.0  \n",
       "904           NaN        8.0  \n",
       "905           NaN       10.0  \n",
       "906          True       10.0  \n",
       "1617         True       12.0  \n",
       "1587          NaN       14.0  \n",
       "1586          NaN       15.0  \n",
       "1588          NaN       17.0  \n",
       "1590          NaN       18.0  \n",
       "907           NaN       21.0  \n",
       "\n",
       "[13 rows x 26 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[metadata['host_id']==23.1].sort_values(by=['collection_date'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b208c0a2-6fbe-4991-9e3e-2620a55e4934",
   "metadata": {},
   "source": [
    "#### The NaN values appear to only be after weaning or when not weaned at all, but not before weaning. (How do we check whether that is really true?)\n",
    "\n",
    "#### If so, we can assign True to each NaN value if metadata['diet_weaning'].sum() >= 1:\n",
    "#### (We discussed this with our tutor and she gave permission to use this assumption)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "81d03df7-b409-4fcd-bffe-027d356eef11",
   "metadata": {},
   "outputs": [],
   "source": [
    "for host in host_numbers:\n",
    "    if metadata['diet_weaning'].sum() >= 1:\n",
    "        metadata['diet_weaning'].fillna(True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcb03a09-93f7-4e1e-9396-e014f1428195",
   "metadata": {},
   "source": [
    "#### Check:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4ba669c3-ae64-4430-8035-e80214443280",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Library Layout</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geo_location_name</th>\n",
       "      <th>geo_latitude</th>\n",
       "      <th>geo_longitude</th>\n",
       "      <th>host_id</th>\n",
       "      <th>age_days</th>\n",
       "      <th>weight_kg</th>\n",
       "      <th>...</th>\n",
       "      <th>birth_length_cm</th>\n",
       "      <th>sex</th>\n",
       "      <th>delivery_mode</th>\n",
       "      <th>zygosity</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>delivery_preterm</th>\n",
       "      <th>diet_milk</th>\n",
       "      <th>diet_weaning</th>\n",
       "      <th>age_months</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>879</th>\n",
       "      <td>ERR1310621</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-05-06 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>47.0</td>\n",
       "      <td>3.487</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>837</th>\n",
       "      <td>ERR1310614</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-07-09 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>111.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1553</th>\n",
       "      <td>ERR1311558</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-08-02 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>135.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>838</th>\n",
       "      <td>ERR1310615</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-09-11 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>175.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1386</th>\n",
       "      <td>ERR1309946</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-10-09 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>203.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>840</th>\n",
       "      <td>ERR1310617</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-11-05 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>230.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>839</th>\n",
       "      <td>ERR1310616</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-11-24 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>249.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>841</th>\n",
       "      <td>ERR1310618</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-01-05 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>291.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1385</th>\n",
       "      <td>ERR1309945</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-02-04 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>322.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1387</th>\n",
       "      <td>ERR1309947</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-02-26 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>343.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>842</th>\n",
       "      <td>ERR1310619</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-03-23 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>369.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1388</th>\n",
       "      <td>ERR1309948</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-05-14 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>420.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1060</th>\n",
       "      <td>ERR1309785</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-06-04 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>441.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>ERR1310620</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-07-01 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>468.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1075</th>\n",
       "      <td>ERR1309786</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-08-12 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>511.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1568</th>\n",
       "      <td>ERR1311559</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-09-09 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>538.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>912</th>\n",
       "      <td>ERR1310622</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-10-03 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>562.0</td>\n",
       "      <td>9.015</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1061</th>\n",
       "      <td>ERR1309787</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-11-23 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>613.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1062</th>\n",
       "      <td>ERR1309788</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-12-09 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>630.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>880</th>\n",
       "      <td>ERR1310623</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-01-05 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>656.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1041</th>\n",
       "      <td>ERR1310505</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-02-06 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>688.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1063</th>\n",
       "      <td>ERR1309789</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-03-01 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>712.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1064</th>\n",
       "      <td>ERR1309790</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-04-07 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>749.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1065</th>\n",
       "      <td>ERR1309791</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-06-04 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>808.0</td>\n",
       "      <td>11.057</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1554</th>\n",
       "      <td>ERR1311560</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-08-17 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>881.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1555</th>\n",
       "      <td>ERR1311561</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-10-20 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>946.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>31.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556</th>\n",
       "      <td>ERR1311562</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-12-11 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>997.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>836</th>\n",
       "      <td>ERR1310613</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2013-02-05 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>16.1</td>\n",
       "      <td>1053.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>42.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id Library Layout      Instrument      collection_date  \\\n",
       "879   ERR1310621         PAIRED  Illumina MiSeq  2010-05-06 00:00:00   \n",
       "837   ERR1310614         PAIRED  Illumina MiSeq  2010-07-09 00:00:00   \n",
       "1553  ERR1311558         PAIRED  Illumina MiSeq  2010-08-02 00:00:00   \n",
       "838   ERR1310615         PAIRED  Illumina MiSeq  2010-09-11 00:00:00   \n",
       "1386  ERR1309946         PAIRED  Illumina MiSeq  2010-10-09 00:00:00   \n",
       "840   ERR1310617         PAIRED  Illumina MiSeq  2010-11-05 00:00:00   \n",
       "839   ERR1310616         PAIRED  Illumina MiSeq  2010-11-24 00:00:00   \n",
       "841   ERR1310618         PAIRED  Illumina MiSeq  2011-01-05 00:00:00   \n",
       "1385  ERR1309945         PAIRED  Illumina MiSeq  2011-02-04 00:00:00   \n",
       "1387  ERR1309947         PAIRED  Illumina MiSeq  2011-02-26 00:00:00   \n",
       "842   ERR1310619         PAIRED  Illumina MiSeq  2011-03-23 00:00:00   \n",
       "1388  ERR1309948         PAIRED  Illumina MiSeq  2011-05-14 00:00:00   \n",
       "1060  ERR1309785         PAIRED  Illumina MiSeq  2011-06-04 00:00:00   \n",
       "844   ERR1310620         PAIRED  Illumina MiSeq  2011-07-01 00:00:00   \n",
       "1075  ERR1309786         PAIRED  Illumina MiSeq  2011-08-12 00:00:00   \n",
       "1568  ERR1311559         PAIRED  Illumina MiSeq  2011-09-09 00:00:00   \n",
       "912   ERR1310622         PAIRED  Illumina MiSeq  2011-10-03 00:00:00   \n",
       "1061  ERR1309787         PAIRED  Illumina MiSeq  2011-11-23 00:00:00   \n",
       "1062  ERR1309788         PAIRED  Illumina MiSeq  2011-12-09 00:00:00   \n",
       "880   ERR1310623         PAIRED  Illumina MiSeq  2012-01-05 00:00:00   \n",
       "1041  ERR1310505         PAIRED  Illumina MiSeq  2012-02-06 00:00:00   \n",
       "1063  ERR1309789         PAIRED  Illumina MiSeq  2012-03-01 00:00:00   \n",
       "1064  ERR1309790         PAIRED  Illumina MiSeq  2012-04-07 00:00:00   \n",
       "1065  ERR1309791         PAIRED  Illumina MiSeq  2012-06-04 00:00:00   \n",
       "1554  ERR1311560         PAIRED  Illumina MiSeq  2012-08-17 00:00:00   \n",
       "1555  ERR1311561         PAIRED  Illumina MiSeq  2012-10-20 00:00:00   \n",
       "1556  ERR1311562         PAIRED  Illumina MiSeq  2012-12-11 00:00:00   \n",
       "836   ERR1310613         PAIRED  Illumina MiSeq  2013-02-05 00:00:00   \n",
       "\n",
       "             geo_location_name  geo_latitude  geo_longitude  host_id  \\\n",
       "879   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "837   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1553  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "838   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1386  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "840   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "839   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "841   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1385  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1387  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "842   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1388  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1060  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "844   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1075  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1568  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "912   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1061  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1062  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "880   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1041  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1063  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1064  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1065  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1554  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1555  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "1556  USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "836   USA, Missouri, St. Louis      38.63699     -90.263794     16.1   \n",
       "\n",
       "      age_days  weight_kg  ...  birth_length_cm     sex  delivery_mode  \\\n",
       "879       47.0      3.487  ...             42.0  female       Cesarean   \n",
       "837      111.0        NaN  ...             42.0  female       Cesarean   \n",
       "1553     135.0        NaN  ...             42.0  female       Cesarean   \n",
       "838      175.0        NaN  ...             42.0  female       Cesarean   \n",
       "1386     203.0        NaN  ...             42.0  female       Cesarean   \n",
       "840      230.0        NaN  ...             42.0  female       Cesarean   \n",
       "839      249.0        NaN  ...             42.0  female       Cesarean   \n",
       "841      291.0        NaN  ...             42.0  female       Cesarean   \n",
       "1385     322.0        NaN  ...             42.0  female       Cesarean   \n",
       "1387     343.0        NaN  ...             42.0  female       Cesarean   \n",
       "842      369.0        NaN  ...             42.0  female       Cesarean   \n",
       "1388     420.0        NaN  ...             42.0  female       Cesarean   \n",
       "1060     441.0        NaN  ...             42.0  female       Cesarean   \n",
       "844      468.0        NaN  ...             42.0  female       Cesarean   \n",
       "1075     511.0        NaN  ...             42.0  female       Cesarean   \n",
       "1568     538.0        NaN  ...             42.0  female       Cesarean   \n",
       "912      562.0      9.015  ...             42.0  female       Cesarean   \n",
       "1061     613.0        NaN  ...             42.0  female       Cesarean   \n",
       "1062     630.0        NaN  ...             42.0  female       Cesarean   \n",
       "880      656.0        NaN  ...             42.0  female       Cesarean   \n",
       "1041     688.0        NaN  ...             42.0  female       Cesarean   \n",
       "1063     712.0        NaN  ...             42.0  female       Cesarean   \n",
       "1064     749.0        NaN  ...             42.0  female       Cesarean   \n",
       "1065     808.0     11.057  ...             42.0  female       Cesarean   \n",
       "1554     881.0        NaN  ...             42.0  female       Cesarean   \n",
       "1555     946.0        NaN  ...             42.0  female       Cesarean   \n",
       "1556     997.0        NaN  ...             42.0  female       Cesarean   \n",
       "836     1053.0        NaN  ...             42.0  female       Cesarean   \n",
       "\n",
       "       zygosity       race     ethnicity  delivery_preterm diet_milk  \\\n",
       "879   Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "837   Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "1553  Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "838   Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "1386  Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "840   Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "839   Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "841   Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "1385  Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "1387  Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "842   Dizygotic  Caucasian  Not Hispanic              True        fd   \n",
       "1388  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1060  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "844   Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1075  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1568  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "912   Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1061  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1062  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "880   Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1041  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1063  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1064  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1065  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1554  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1555  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "1556  Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "836   Dizygotic  Caucasian  Not Hispanic              True       NaN   \n",
       "\n",
       "     diet_weaning age_months  \n",
       "879         False        2.0  \n",
       "837          True        4.0  \n",
       "1553        False        4.0  \n",
       "838          True        6.0  \n",
       "1386         True        7.0  \n",
       "840          True        8.0  \n",
       "839          True        8.0  \n",
       "841          True       10.0  \n",
       "1385         True       11.0  \n",
       "1387         True       11.0  \n",
       "842          True       12.0  \n",
       "1388          NaN       14.0  \n",
       "1060          NaN       14.0  \n",
       "844           NaN       15.0  \n",
       "1075          NaN       17.0  \n",
       "1568          NaN       18.0  \n",
       "912           NaN       18.0  \n",
       "1061          NaN       20.0  \n",
       "1062          NaN       21.0  \n",
       "880           NaN       22.0  \n",
       "1041          NaN       23.0  \n",
       "1063          NaN       23.0  \n",
       "1064          NaN       25.0  \n",
       "1065          NaN       27.0  \n",
       "1554          NaN       29.0  \n",
       "1555          NaN       31.0  \n",
       "1556          NaN       33.0  \n",
       "836           NaN       35.0  \n",
       "\n",
       "[28 rows x 26 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_hosts[16.1].sort_values(by=['collection_date'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "349f0007-b377-41b8-961f-a3e65209cd71",
   "metadata": {},
   "source": [
    "### 3. Get table for each twin pair and each stage"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8d4c164-14b4-4f56-9425-f993050618db",
   "metadata": {},
   "source": [
    "#### Pair the twins:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "84e76a8c-3368-4e77-9fe4-acaff565ef03",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "pair_numbers = list(dict.fromkeys(host_numbers.round(0)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1596b16c-54ea-4c0e-9601-a4f2b7ea225b",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_pairs = dict()\n",
    "for pair in pair_numbers: #loop through all unique pair ids\n",
    "    a = metadata[metadata['host_id'] == pair+0.1]\n",
    "    b = metadata[metadata['host_id'] == pair+0.2]\n",
    "    all_pairs[pair] = pd.concat([a,b])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b054879d-c7ef-4301-90f9-906a66c4c2e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#why was this necessary in the previous dictionary loop?\n",
    "#new_name = 'df_pair_'+str(pair)\n",
    "#locals()[new_name] = metadata[metadata['host_id']==host]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b455b4fd-152a-414a-81ed-5c1e6e454a12",
   "metadata": {},
   "source": [
    "#### Get tables for each pair and each stage:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e4f5dc56-469b-4d6d-a58b-2a754756ee58",
   "metadata": {},
   "outputs": [],
   "source": [
    "stages = metadata['diet_milk'].dropna().unique()\n",
    "stages\n",
    "\n",
    "all_stages = dict()\n",
    "\n",
    "for stage in stages:\n",
    "    new_name = 'df_stage_'+str(stage)\n",
    "    all_stages[stage] = metadata[metadata['diet_milk'] == stage]\n",
    "    locals()[new_name] = metadata[metadata['diet_milk'] == stage]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26af4dcc-926a-4201-b447-30addd193242",
   "metadata": {},
   "source": [
    "#### Check:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "4bfaf132-2312-443a-b62c-afc91c12e758",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Library Layout</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geo_location_name</th>\n",
       "      <th>geo_latitude</th>\n",
       "      <th>geo_longitude</th>\n",
       "      <th>host_id</th>\n",
       "      <th>age_days</th>\n",
       "      <th>weight_kg</th>\n",
       "      <th>...</th>\n",
       "      <th>birth_length_cm</th>\n",
       "      <th>sex</th>\n",
       "      <th>delivery_mode</th>\n",
       "      <th>zygosity</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>delivery_preterm</th>\n",
       "      <th>diet_milk</th>\n",
       "      <th>diet_weaning</th>\n",
       "      <th>age_months</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR1314182</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-11-11 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>42.1</td>\n",
       "      <td>232.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>47.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR1314183</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-12-11 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>27.2</td>\n",
       "      <td>192.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>45.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ERR1314190</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-02-12 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>324.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ERR1314191</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-02-12 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.2</td>\n",
       "      <td>325.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ERR1314198</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-04-12 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>47.1</td>\n",
       "      <td>305.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>46.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>African-American</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1661</th>\n",
       "      <td>ERR1310702</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-04-12 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>25.2</td>\n",
       "      <td>335.0</td>\n",
       "      <td>9.696</td>\n",
       "      <td>...</td>\n",
       "      <td>48.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean_emergency</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>False</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1664</th>\n",
       "      <td>ERR1310705</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-07-31 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>26.1</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>False</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665</th>\n",
       "      <td>ERR1310707</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-02-07 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>27.1</td>\n",
       "      <td>250.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>45.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1666</th>\n",
       "      <td>ERR1310708</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-04-09 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>27.1</td>\n",
       "      <td>310.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>45.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Cesarean</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1668</th>\n",
       "      <td>ERR1310706</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-07-31 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>26.2</td>\n",
       "      <td>73.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Dizygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>False</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>588 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id Library Layout      Instrument      collection_date  \\\n",
       "0     ERR1314182         PAIRED  Illumina MiSeq  2011-11-11 00:00:00   \n",
       "1     ERR1314183         PAIRED  Illumina MiSeq  2010-12-11 00:00:00   \n",
       "8     ERR1314190         PAIRED  Illumina MiSeq  2012-02-12 00:00:00   \n",
       "9     ERR1314191         PAIRED  Illumina MiSeq  2012-02-12 00:00:00   \n",
       "13    ERR1314198         PAIRED  Illumina MiSeq  2012-04-12 00:00:00   \n",
       "...          ...            ...             ...                  ...   \n",
       "1661  ERR1310702         PAIRED  Illumina MiSeq  2011-04-12 00:00:00   \n",
       "1664  ERR1310705         PAIRED  Illumina MiSeq  2010-07-31 00:00:00   \n",
       "1665  ERR1310707         PAIRED  Illumina MiSeq  2011-02-07 00:00:00   \n",
       "1666  ERR1310708         PAIRED  Illumina MiSeq  2011-04-09 00:00:00   \n",
       "1668  ERR1310706         PAIRED  Illumina MiSeq  2010-07-31 00:00:00   \n",
       "\n",
       "             geo_location_name  geo_latitude  geo_longitude  host_id  \\\n",
       "0     USA, Missouri, St. Louis      38.63699     -90.263794     42.1   \n",
       "1     USA, Missouri, St. Louis      38.63699     -90.263794     27.2   \n",
       "8     USA, Missouri, St. Louis      38.63699     -90.263794     40.1   \n",
       "9     USA, Missouri, St. Louis      38.63699     -90.263794     40.2   \n",
       "13    USA, Missouri, St. Louis      38.63699     -90.263794     47.1   \n",
       "...                        ...           ...            ...      ...   \n",
       "1661  USA, Missouri, St. Louis      38.63699     -90.263794     25.2   \n",
       "1664  USA, Missouri, St. Louis      38.63699     -90.263794     26.1   \n",
       "1665  USA, Missouri, St. Louis      38.63699     -90.263794     27.1   \n",
       "1666  USA, Missouri, St. Louis      38.63699     -90.263794     27.1   \n",
       "1668  USA, Missouri, St. Louis      38.63699     -90.263794     26.2   \n",
       "\n",
       "      age_days  weight_kg  ...  birth_length_cm     sex       delivery_mode  \\\n",
       "0        232.0        NaN  ...             47.0    male            Cesarean   \n",
       "1        192.0        NaN  ...             45.0  female            Cesarean   \n",
       "8        324.0        NaN  ...             49.0  female             Vaginal   \n",
       "9        325.0        NaN  ...             44.0  female             Vaginal   \n",
       "13       305.0        NaN  ...             46.0    male             Vaginal   \n",
       "...        ...        ...  ...              ...     ...                 ...   \n",
       "1661     335.0      9.696  ...             48.0  female  Cesarean_emergency   \n",
       "1664      74.0        NaN  ...             49.0  female             Vaginal   \n",
       "1665     250.0        NaN  ...             45.0  female            Cesarean   \n",
       "1666     310.0        NaN  ...             45.0  female            Cesarean   \n",
       "1668      73.0        NaN  ...             49.0  female             Vaginal   \n",
       "\n",
       "         zygosity              race     ethnicity  delivery_preterm diet_milk  \\\n",
       "0     Monozygotic         Caucasian  Not Hispanic              True        fd   \n",
       "1       Dizygotic         Caucasian      Hispanic              True        fd   \n",
       "8     Monozygotic         Caucasian  Not Hispanic              True        fd   \n",
       "9     Monozygotic         Caucasian  Not Hispanic              True        fd   \n",
       "13    Monozygotic  African-American  Not Hispanic              True        fd   \n",
       "...           ...               ...           ...               ...       ...   \n",
       "1661  Monozygotic         Caucasian      Hispanic             False        fd   \n",
       "1664    Dizygotic         Caucasian  Not Hispanic             False        fd   \n",
       "1665    Dizygotic         Caucasian      Hispanic              True        fd   \n",
       "1666    Dizygotic         Caucasian      Hispanic              True        fd   \n",
       "1668    Dizygotic         Caucasian  Not Hispanic             False        fd   \n",
       "\n",
       "     diet_weaning age_months  \n",
       "0            True        8.0  \n",
       "1            True        6.0  \n",
       "8            True       11.0  \n",
       "9            True       11.0  \n",
       "13           True       10.0  \n",
       "...           ...        ...  \n",
       "1661         True       11.0  \n",
       "1664        False        2.0  \n",
       "1665         True        8.0  \n",
       "1666         True       10.0  \n",
       "1668        False        2.0  \n",
       "\n",
       "[588 rows x 26 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_stages['fd']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee856925-cc5b-4d6b-852f-bed1593b6399",
   "metadata": {},
   "source": [
    "#### In order to generate a table that contains values of two dictionaries with separate conditions, we can merge them like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "5e301d6d-fc71-4ed3-84bd-8270f2208b10",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Library Layout</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geo_location_name</th>\n",
       "      <th>geo_latitude</th>\n",
       "      <th>geo_longitude</th>\n",
       "      <th>host_id</th>\n",
       "      <th>age_days</th>\n",
       "      <th>weight_kg</th>\n",
       "      <th>...</th>\n",
       "      <th>birth_length_cm</th>\n",
       "      <th>sex</th>\n",
       "      <th>delivery_mode</th>\n",
       "      <th>zygosity</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>delivery_preterm</th>\n",
       "      <th>diet_milk</th>\n",
       "      <th>diet_weaning</th>\n",
       "      <th>age_months</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR1314190</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-02-12 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>324.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR1314191</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-02-12 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.2</td>\n",
       "      <td>325.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR1314250</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-10-12 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>200.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR1314054</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-11-09 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>229.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR1314055</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-11-09 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.2</td>\n",
       "      <td>230.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ERR1314557</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-09-14 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.2</td>\n",
       "      <td>173.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ERR1314295</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-09-13 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>172.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ERR1314716</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-08-15 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>144.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ERR1314717</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-08-15 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.2</td>\n",
       "      <td>143.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ERR1315635</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-07-29 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.2</td>\n",
       "      <td>126.0</td>\n",
       "      <td>5.273</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ERR1315536</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-07-27 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>124.0</td>\n",
       "      <td>6.974</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ERR1315527</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-05-27 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.2</td>\n",
       "      <td>64.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ERR1315526</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-05-27 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>63.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ERR1315590</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-06-28 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.2</td>\n",
       "      <td>96.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ERR1315589</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-06-28 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ERR1313847</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-01-08 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>289.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ERR1313848</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2012-01-08 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.2</td>\n",
       "      <td>289.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ERR1313952</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-12-08 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.1</td>\n",
       "      <td>258.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ERR1313954</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2011-12-08 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>40.2</td>\n",
       "      <td>258.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>female</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            id Library Layout      Instrument      collection_date  \\\n",
       "0   ERR1314190         PAIRED  Illumina MiSeq  2012-02-12 00:00:00   \n",
       "1   ERR1314191         PAIRED  Illumina MiSeq  2012-02-12 00:00:00   \n",
       "2   ERR1314250         PAIRED  Illumina MiSeq  2011-10-12 00:00:00   \n",
       "3   ERR1314054         PAIRED  Illumina MiSeq  2011-11-09 00:00:00   \n",
       "4   ERR1314055         PAIRED  Illumina MiSeq  2011-11-09 00:00:00   \n",
       "5   ERR1314557         PAIRED  Illumina MiSeq  2011-09-14 00:00:00   \n",
       "6   ERR1314295         PAIRED  Illumina MiSeq  2011-09-13 00:00:00   \n",
       "7   ERR1314716         PAIRED  Illumina MiSeq  2011-08-15 00:00:00   \n",
       "8   ERR1314717         PAIRED  Illumina MiSeq  2011-08-15 00:00:00   \n",
       "9   ERR1315635         PAIRED  Illumina MiSeq  2011-07-29 00:00:00   \n",
       "10  ERR1315536         PAIRED  Illumina MiSeq  2011-07-27 00:00:00   \n",
       "11  ERR1315527         PAIRED  Illumina MiSeq  2011-05-27 00:00:00   \n",
       "12  ERR1315526         PAIRED  Illumina MiSeq  2011-05-27 00:00:00   \n",
       "13  ERR1315590         PAIRED  Illumina MiSeq  2011-06-28 00:00:00   \n",
       "14  ERR1315589         PAIRED  Illumina MiSeq  2011-06-28 00:00:00   \n",
       "15  ERR1313847         PAIRED  Illumina MiSeq  2012-01-08 00:00:00   \n",
       "16  ERR1313848         PAIRED  Illumina MiSeq  2012-01-08 00:00:00   \n",
       "17  ERR1313952         PAIRED  Illumina MiSeq  2011-12-08 00:00:00   \n",
       "18  ERR1313954         PAIRED  Illumina MiSeq  2011-12-08 00:00:00   \n",
       "\n",
       "           geo_location_name  geo_latitude  geo_longitude  host_id  age_days  \\\n",
       "0   USA, Missouri, St. Louis      38.63699     -90.263794     40.1     324.0   \n",
       "1   USA, Missouri, St. Louis      38.63699     -90.263794     40.2     325.0   \n",
       "2   USA, Missouri, St. Louis      38.63699     -90.263794     40.1     200.0   \n",
       "3   USA, Missouri, St. Louis      38.63699     -90.263794     40.1     229.0   \n",
       "4   USA, Missouri, St. Louis      38.63699     -90.263794     40.2     230.0   \n",
       "5   USA, Missouri, St. Louis      38.63699     -90.263794     40.2     173.0   \n",
       "6   USA, Missouri, St. Louis      38.63699     -90.263794     40.1     172.0   \n",
       "7   USA, Missouri, St. Louis      38.63699     -90.263794     40.1     144.0   \n",
       "8   USA, Missouri, St. Louis      38.63699     -90.263794     40.2     143.0   \n",
       "9   USA, Missouri, St. Louis      38.63699     -90.263794     40.2     126.0   \n",
       "10  USA, Missouri, St. Louis      38.63699     -90.263794     40.1     124.0   \n",
       "11  USA, Missouri, St. Louis      38.63699     -90.263794     40.2      64.0   \n",
       "12  USA, Missouri, St. Louis      38.63699     -90.263794     40.1      63.0   \n",
       "13  USA, Missouri, St. Louis      38.63699     -90.263794     40.2      96.0   \n",
       "14  USA, Missouri, St. Louis      38.63699     -90.263794     40.1      95.0   \n",
       "15  USA, Missouri, St. Louis      38.63699     -90.263794     40.1     289.0   \n",
       "16  USA, Missouri, St. Louis      38.63699     -90.263794     40.2     289.0   \n",
       "17  USA, Missouri, St. Louis      38.63699     -90.263794     40.1     258.0   \n",
       "18  USA, Missouri, St. Louis      38.63699     -90.263794     40.2     258.0   \n",
       "\n",
       "    weight_kg  ...  birth_length_cm     sex  delivery_mode     zygosity  \\\n",
       "0         NaN  ...             49.0  female        Vaginal  Monozygotic   \n",
       "1         NaN  ...             44.0  female        Vaginal  Monozygotic   \n",
       "2         NaN  ...             49.0  female        Vaginal  Monozygotic   \n",
       "3         NaN  ...             49.0  female        Vaginal  Monozygotic   \n",
       "4         NaN  ...             44.0  female        Vaginal  Monozygotic   \n",
       "5         NaN  ...             44.0  female        Vaginal  Monozygotic   \n",
       "6         NaN  ...             49.0  female        Vaginal  Monozygotic   \n",
       "7         NaN  ...             49.0  female        Vaginal  Monozygotic   \n",
       "8         NaN  ...             44.0  female        Vaginal  Monozygotic   \n",
       "9       5.273  ...             44.0  female        Vaginal  Monozygotic   \n",
       "10      6.974  ...             49.0  female        Vaginal  Monozygotic   \n",
       "11        NaN  ...             44.0  female        Vaginal  Monozygotic   \n",
       "12        NaN  ...             49.0  female        Vaginal  Monozygotic   \n",
       "13        NaN  ...             44.0  female        Vaginal  Monozygotic   \n",
       "14        NaN  ...             49.0  female        Vaginal  Monozygotic   \n",
       "15        NaN  ...             49.0  female        Vaginal  Monozygotic   \n",
       "16        NaN  ...             44.0  female        Vaginal  Monozygotic   \n",
       "17        NaN  ...             49.0  female        Vaginal  Monozygotic   \n",
       "18        NaN  ...             44.0  female        Vaginal  Monozygotic   \n",
       "\n",
       "         race     ethnicity  delivery_preterm diet_milk diet_weaning  \\\n",
       "0   Caucasian  Not Hispanic              True        fd         True   \n",
       "1   Caucasian  Not Hispanic              True        fd         True   \n",
       "2   Caucasian  Not Hispanic              True        fd         True   \n",
       "3   Caucasian  Not Hispanic              True        fd         True   \n",
       "4   Caucasian  Not Hispanic              True        fd         True   \n",
       "5   Caucasian  Not Hispanic              True        fd         True   \n",
       "6   Caucasian  Not Hispanic              True        fd         True   \n",
       "7   Caucasian  Not Hispanic              True        fd         True   \n",
       "8   Caucasian  Not Hispanic              True        fd         True   \n",
       "9   Caucasian  Not Hispanic              True        fd        False   \n",
       "10  Caucasian  Not Hispanic              True        fd        False   \n",
       "11  Caucasian  Not Hispanic              True        fd        False   \n",
       "12  Caucasian  Not Hispanic              True        fd        False   \n",
       "13  Caucasian  Not Hispanic              True        fd         True   \n",
       "14  Caucasian  Not Hispanic              True        fd        False   \n",
       "15  Caucasian  Not Hispanic              True        fd         True   \n",
       "16  Caucasian  Not Hispanic              True        fd         True   \n",
       "17  Caucasian  Not Hispanic              True        fd         True   \n",
       "18  Caucasian  Not Hispanic              True        fd         True   \n",
       "\n",
       "   age_months  \n",
       "0        11.0  \n",
       "1        11.0  \n",
       "2         7.0  \n",
       "3         8.0  \n",
       "4         8.0  \n",
       "5         6.0  \n",
       "6         6.0  \n",
       "7         5.0  \n",
       "8         5.0  \n",
       "9         4.0  \n",
       "10        4.0  \n",
       "11        2.0  \n",
       "12        2.0  \n",
       "13        3.0  \n",
       "14        3.0  \n",
       "15        9.0  \n",
       "16        9.0  \n",
       "17        8.0  \n",
       "18        8.0  \n",
       "\n",
       "[19 rows x 26 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_stages['fd'].merge(all_pairs[40], how = 'inner')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5906d8aa-776a-4e0d-9661-3ff83597ddc9",
   "metadata": {},
   "source": [
    "#### That way we can generate a new dictionary with which we can recall all pairs with a certain diet:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a1ca7eb7-72f0-434b-83a3-63cb87f94017",
   "metadata": {},
   "outputs": [],
   "source": [
    "pair_stages = dict()\n",
    "\n",
    "for pair in pair_numbers:\n",
    "    for stage in stages:\n",
    "        pair_stages[pair,stage] = all_stages[stage].merge(all_pairs[pair], how = 'inner')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd886cb1-65fc-4a45-b39a-9e439adf6859",
   "metadata": {},
   "source": [
    "#### Check:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f474457a-fc66-45c5-8cbb-f8e4f6d5bacc",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Library Layout</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>collection_date</th>\n",
       "      <th>geo_location_name</th>\n",
       "      <th>geo_latitude</th>\n",
       "      <th>geo_longitude</th>\n",
       "      <th>host_id</th>\n",
       "      <th>age_days</th>\n",
       "      <th>weight_kg</th>\n",
       "      <th>...</th>\n",
       "      <th>birth_length_cm</th>\n",
       "      <th>sex</th>\n",
       "      <th>delivery_mode</th>\n",
       "      <th>zygosity</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>delivery_preterm</th>\n",
       "      <th>diet_milk</th>\n",
       "      <th>diet_weaning</th>\n",
       "      <th>age_months</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR1314203</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-06-12 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.1</td>\n",
       "      <td>151.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>48.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR1314448</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-12-13 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.2</td>\n",
       "      <td>336.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR1314828</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-04-17 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.1</td>\n",
       "      <td>95.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>48.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR1314888</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-04-18 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.2</td>\n",
       "      <td>96.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR1314844</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-06-17 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.2</td>\n",
       "      <td>156.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ERR1314860</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-11-17 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.1</td>\n",
       "      <td>309.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>48.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ERR1314640</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-05-15 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.2</td>\n",
       "      <td>123.0</td>\n",
       "      <td>6.606</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ERR1315089</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-12-21 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.1</td>\n",
       "      <td>344.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>48.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ERR1315188</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-09-24 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.2</td>\n",
       "      <td>256.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ERR1315190</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-10-24 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.2</td>\n",
       "      <td>286.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ERR1314998</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-03-20 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.1</td>\n",
       "      <td>68.0</td>\n",
       "      <td>4.635</td>\n",
       "      <td>...</td>\n",
       "      <td>48.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ERR1314999</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-03-20 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.2</td>\n",
       "      <td>68.0</td>\n",
       "      <td>4.720</td>\n",
       "      <td>...</td>\n",
       "      <td>49.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>False</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ERR1315071</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>2010-09-21 00:00:00</td>\n",
       "      <td>USA, Missouri, St. Louis</td>\n",
       "      <td>38.63699</td>\n",
       "      <td>-90.263794</td>\n",
       "      <td>3.1</td>\n",
       "      <td>253.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>48.0</td>\n",
       "      <td>male</td>\n",
       "      <td>Vaginal</td>\n",
       "      <td>Monozygotic</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>Not Hispanic</td>\n",
       "      <td>True</td>\n",
       "      <td>fd</td>\n",
       "      <td>True</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            id Library Layout      Instrument      collection_date  \\\n",
       "0   ERR1314203         PAIRED  Illumina MiSeq  2010-06-12 00:00:00   \n",
       "1   ERR1314448         PAIRED  Illumina MiSeq  2010-12-13 00:00:00   \n",
       "2   ERR1314828         PAIRED  Illumina MiSeq  2010-04-17 00:00:00   \n",
       "3   ERR1314888         PAIRED  Illumina MiSeq  2010-04-18 00:00:00   \n",
       "4   ERR1314844         PAIRED  Illumina MiSeq  2010-06-17 00:00:00   \n",
       "5   ERR1314860         PAIRED  Illumina MiSeq  2010-11-17 00:00:00   \n",
       "6   ERR1314640         PAIRED  Illumina MiSeq  2010-05-15 00:00:00   \n",
       "7   ERR1315089         PAIRED  Illumina MiSeq  2010-12-21 00:00:00   \n",
       "8   ERR1315188         PAIRED  Illumina MiSeq  2010-09-24 00:00:00   \n",
       "9   ERR1315190         PAIRED  Illumina MiSeq  2010-10-24 00:00:00   \n",
       "10  ERR1314998         PAIRED  Illumina MiSeq  2010-03-20 00:00:00   \n",
       "11  ERR1314999         PAIRED  Illumina MiSeq  2010-03-20 00:00:00   \n",
       "12  ERR1315071         PAIRED  Illumina MiSeq  2010-09-21 00:00:00   \n",
       "\n",
       "           geo_location_name  geo_latitude  geo_longitude  host_id  age_days  \\\n",
       "0   USA, Missouri, St. Louis      38.63699     -90.263794      3.1     151.0   \n",
       "1   USA, Missouri, St. Louis      38.63699     -90.263794      3.2     336.0   \n",
       "2   USA, Missouri, St. Louis      38.63699     -90.263794      3.1      95.0   \n",
       "3   USA, Missouri, St. Louis      38.63699     -90.263794      3.2      96.0   \n",
       "4   USA, Missouri, St. Louis      38.63699     -90.263794      3.2     156.0   \n",
       "5   USA, Missouri, St. Louis      38.63699     -90.263794      3.1     309.0   \n",
       "6   USA, Missouri, St. Louis      38.63699     -90.263794      3.2     123.0   \n",
       "7   USA, Missouri, St. Louis      38.63699     -90.263794      3.1     344.0   \n",
       "8   USA, Missouri, St. Louis      38.63699     -90.263794      3.2     256.0   \n",
       "9   USA, Missouri, St. Louis      38.63699     -90.263794      3.2     286.0   \n",
       "10  USA, Missouri, St. Louis      38.63699     -90.263794      3.1      68.0   \n",
       "11  USA, Missouri, St. Louis      38.63699     -90.263794      3.2      68.0   \n",
       "12  USA, Missouri, St. Louis      38.63699     -90.263794      3.1     253.0   \n",
       "\n",
       "    weight_kg  ...  birth_length_cm   sex  delivery_mode     zygosity  \\\n",
       "0         NaN  ...             48.0  male        Vaginal  Monozygotic   \n",
       "1         NaN  ...             49.0  male        Vaginal  Monozygotic   \n",
       "2         NaN  ...             48.0  male        Vaginal  Monozygotic   \n",
       "3         NaN  ...             49.0  male        Vaginal  Monozygotic   \n",
       "4         NaN  ...             49.0  male        Vaginal  Monozygotic   \n",
       "5         NaN  ...             48.0  male        Vaginal  Monozygotic   \n",
       "6       6.606  ...             49.0  male        Vaginal  Monozygotic   \n",
       "7         NaN  ...             48.0  male        Vaginal  Monozygotic   \n",
       "8         NaN  ...             49.0  male        Vaginal  Monozygotic   \n",
       "9         NaN  ...             49.0  male        Vaginal  Monozygotic   \n",
       "10      4.635  ...             48.0  male        Vaginal  Monozygotic   \n",
       "11      4.720  ...             49.0  male        Vaginal  Monozygotic   \n",
       "12        NaN  ...             48.0  male        Vaginal  Monozygotic   \n",
       "\n",
       "         race     ethnicity  delivery_preterm diet_milk diet_weaning  \\\n",
       "0   Caucasian  Not Hispanic              True        fd         True   \n",
       "1   Caucasian  Not Hispanic              True        fd         True   \n",
       "2   Caucasian  Not Hispanic              True        fd        False   \n",
       "3   Caucasian  Not Hispanic              True        fd        False   \n",
       "4   Caucasian  Not Hispanic              True        fd         True   \n",
       "5   Caucasian  Not Hispanic              True        fd         True   \n",
       "6   Caucasian  Not Hispanic              True        fd         True   \n",
       "7   Caucasian  Not Hispanic              True        fd         True   \n",
       "8   Caucasian  Not Hispanic              True        fd         True   \n",
       "9   Caucasian  Not Hispanic              True        fd         True   \n",
       "10  Caucasian  Not Hispanic              True        fd        False   \n",
       "11  Caucasian  Not Hispanic              True        fd        False   \n",
       "12  Caucasian  Not Hispanic              True        fd         True   \n",
       "\n",
       "   age_months  \n",
       "0         5.0  \n",
       "1        11.0  \n",
       "2         3.0  \n",
       "3         3.0  \n",
       "4         5.0  \n",
       "5        10.0  \n",
       "6         4.0  \n",
       "7        11.0  \n",
       "8         8.0  \n",
       "9         9.0  \n",
       "10        2.0  \n",
       "11        2.0  \n",
       "12        8.0  \n",
       "\n",
       "[13 rows x 26 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pair_stages[3, 'fd']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66f565db-c2ab-492c-8e40-45dd17c1598e",
   "metadata": {},
   "source": [
    "### 4. Filter feature table according to metadata table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4de5a0db-68a1-47fc-b685-572609f5ae38",
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime feature-table filter-samples --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9ca83a76-511d-4ab9-9ddf-28f5707373b7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "meta_wean = pd.read_csv(data_dir + '/metadata_only_before_after_weaning.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5335762d-29d2-4df0-a12f-f832c5f15106",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_wean.to_csv('meta_wean.tsv', sep='\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "890b6984-2e38-42e0-8785-9d0a249cba27",
   "metadata": {},
   "source": [
    "#### Weaned:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5a5db615-449a-4281-a140-9d63f722b2c2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_42.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_27.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_28.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_39.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_8.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_29.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_40.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_35.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_47.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_4.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_3.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_30.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_36.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_6.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_33.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_43.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_44.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_45.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_5.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_37.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_46.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_48.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_32.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_12.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_13.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_14.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_10.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_15.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_16.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_25.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_26.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_11.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_2.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_20.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_21.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_23.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_19.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_17.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_18.0_weaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_weaned/pair_24.0_weaned.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "for pair in pair_numbers:\n",
    "    pair1, pair2 = pair + 0.1, pair + 0.2\n",
    "     \n",
    "    ! qiime feature-table filter-samples \\\n",
    "        --i-table $data_dir/phylogeny_filtered_table.qza \\\n",
    "        --m-metadata-file $data_dir/meta_wean.tsv \\\n",
    "        --p-where \"[diet_weaning]='weaned' and ([host_id]=$pair1 or [host_id]=$pair2)\" \\\n",
    "        --o-filtered-table $data_dir/pairs_weaned/pair_\"$pair\"_weaned.qza"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40777c82-f64c-4f8f-bef0-a1959a3db7bd",
   "metadata": {},
   "source": [
    "#### Not Weaned:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f15e2a2a-dacb-4568-8969-ba56198ec9b3",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_42.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_27.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_28.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_39.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_8.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_29.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_40.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_35.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_47.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_4.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_3.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_30.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_36.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_6.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_33.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_43.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_44.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_45.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_5.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_37.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_46.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_48.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_32.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_12.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_13.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_14.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_10.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_15.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_16.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_25.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_26.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_11.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_2.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_20.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_21.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_23.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_19.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_17.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_18.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Frequency] to: data/pairs_notweaned/pair_24.0_notweaned.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "for pair in pair_numbers:\n",
    "    pair1, pair2 = pair + 0.1, pair + 0.2\n",
    "    \n",
    "    ! qiime feature-table filter-samples \\\n",
    "    --i-table $data_dir/phylogeny_filtered_table.qza \\\n",
    "    --m-metadata-file $data_dir/meta_wean.tsv \\\n",
    "    --p-where \"[diet_weaning]='not weaning' and ([host_id]=$pair1 or [host_id]=$pair2)\" \\\n",
    "    --o-filtered-table $data_dir/pairs_notweaned/pair_\"$pair\"_notweaned.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "68fb9915-5f1e-4462-a295-c11577a698b1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: \u001b[94mqiime feature-table filter-samples\u001b[0m [OPTIONS]\n",
      "\n",
      "  Filter samples from table based on frequency and/or metadata. Any features\n",
      "  with a frequency of zero after sample filtering will also be removed. See\n",
      "  the filtering tutorial on https://docs.qiime2.org for additional details.\n",
      "\n",
      "\u001b[1mInputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--i-table\u001b[0m ARTIFACT \u001b[32mFeatureTable[Frequency¹ | RelativeFrequency² |\u001b[0m\n",
      "    \u001b[32mPresenceAbsence³ | Composition⁴]\u001b[0m\n",
      "                       The feature table from which samples should be\n",
      "                       filtered.                                    \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mParameters\u001b[0m:\n",
      "  \u001b[94m--p-min-frequency\u001b[0m INTEGER\n",
      "                       The minimum total frequency that a sample must have to\n",
      "                       be retained.                               \u001b[35m[default: 0]\u001b[0m\n",
      "  \u001b[94m--p-max-frequency\u001b[0m INTEGER\n",
      "                       The maximum total frequency that a sample can have to\n",
      "                       be retained. If no value is provided this will default\n",
      "                       to infinity (i.e., no maximum frequency filter will be\n",
      "                       applied).                                    \u001b[35m[optional]\u001b[0m\n",
      "  \u001b[94m--p-min-features\u001b[0m INTEGER\n",
      "                       The minimum number of features that a sample must have\n",
      "                       to be retained.                            \u001b[35m[default: 0]\u001b[0m\n",
      "  \u001b[94m--p-max-features\u001b[0m INTEGER\n",
      "                       The maximum number of features that a sample can have\n",
      "                       to be retained. If no value is provided this will\n",
      "                       default to infinity (i.e., no maximum feature filter\n",
      "                       will be applied).                            \u001b[35m[optional]\u001b[0m\n",
      "  \u001b[94m--m-metadata-file\u001b[0m METADATA...\n",
      "    (multiple          Sample metadata used with `where` parameter when\n",
      "     arguments will    selecting samples to retain, or with `\u001b[4mexclude-ids\u001b[0m` when\n",
      "     be merged)        selecting samples to discard.                \u001b[35m[optional]\u001b[0m\n",
      "  \u001b[94m--p-where\u001b[0m TEXT       SQLite WHERE clause specifying sample metadata\n",
      "                       criteria that must be met to be included in the\n",
      "                       filtered feature table. If not provided, all samples in\n",
      "                       `metadata` that are also in the feature table will be\n",
      "                       retained.                                    \u001b[35m[optional]\u001b[0m\n",
      "  \u001b[94m--p-exclude-ids\u001b[0m / \u001b[94m--p-no-exclude-ids\u001b[0m\n",
      "                       If true, the samples selected by `metadata` or `where`\n",
      "                       parameters will be excluded from the filtered table\n",
      "                       instead of being retained.             \u001b[35m[default: False]\u001b[0m\n",
      "  \u001b[94m--p-filter-empty-features\u001b[0m / \u001b[94m--p-no-filter-empty-features\u001b[0m\n",
      "                       If true, features which are not present in any\n",
      "                       retained samples are dropped.           \u001b[35m[default: True]\u001b[0m\n",
      "\u001b[1mOutputs\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--o-filtered-table\u001b[0m ARTIFACT \u001b[32mFeatureTable[Frequency¹ | RelativeFrequency²\u001b[0m\n",
      "    \u001b[32m| PresenceAbsence³ | Composition⁴]\u001b[0m\n",
      "                       The resulting feature table filtered by sample.\n",
      "                                                                    \u001b[35m[required]\u001b[0m\n",
      "\u001b[1mMiscellaneous\u001b[0m:\n",
      "  \u001b[94m--output-dir\u001b[0m PATH    Output unspecified results to a directory\n",
      "  \u001b[94m--verbose\u001b[0m / \u001b[94m--quiet\u001b[0m  Display verbose output to stdout and/or stderr during\n",
      "                       execution of this action. Or silence output if\n",
      "                       execution is successful (silence is golden).\n",
      "  \u001b[94m--example-data\u001b[0m PATH  Write example data and exit.\n",
      "  \u001b[94m--citations\u001b[0m          Show citations and exit.\n",
      "  \u001b[94m--help\u001b[0m               Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    " ! qiime feature-table filter-samples --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63f5e3a4-109d-467a-b0d9-2cde8ef89263",
   "metadata": {},
   "source": [
    "### 5. Find F values for twin column with ANCOM showing differences between individuals"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b9b060f-85bf-48ca-b918-137e9201fcc2",
   "metadata": {},
   "source": [
    "#### For the ANCOM I need the host ID column as categorical instead of numeric."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2808705a-ccec-4605-a4b3-a1481d66270f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: \u001b[94mqiime tools cast-metadata\u001b[0m [OPTIONS] METADATA...\n",
      "\n",
      "  Designate metadata column types. Supported column types are as follows:\n",
      "  categorical, numeric. Providing multiple file paths to this command will\n",
      "  merge the metadata.\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--cast\u001b[0m COLUMN:TYPE  Parameter for each metadata column that should be cast\n",
      "                      as a specified column type (supported types are as\n",
      "                      follows: categorical, numeric). The required formatting\n",
      "                      for this parameter is --cast COLUMN:TYPE, repeated for\n",
      "                      each column and the associated column type it should be\n",
      "                      cast to in the output.                        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--ignore-extra\u001b[0m      If this flag is enabled, cast parameters that do not\n",
      "                      correspond to any of the column names within the\n",
      "                      provided metadata will be ignored.\n",
      "  \u001b[94m--error-on-missing\u001b[0m  If this flag is enabled, failing to include cast\n",
      "                      parameters for all columns in the provided metadata will\n",
      "                      result in an error.\n",
      "  \u001b[94m--output-file\u001b[0m FILE  Path to file where the modified metadata should be\n",
      "                      written to.\n",
      "  \u001b[94m--help\u001b[0m              Show this message and exit.\n"
     ]
    }
   ],
   "source": [
    "! qiime tools cast-metadata --help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "afdb8648-33e1-42df-a262-95dddbdb1844",
   "metadata": {},
   "outputs": [],
   "source": [
    "! qiime tools cast-metadata $data_dir/meta_wean.tsv \\\n",
    "--cast host_id:categorical \\\n",
    "--output-file $data_dir/meta_wean2.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d7afb2d-4f3c-4aa4-9409-838ed5f544b3",
   "metadata": {},
   "source": [
    "#### For Weaned:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "af40efd9-e77f-4a3c-b76d-b21e85e60d2e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_42.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_27.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_28.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_39.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_8.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_29.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_40.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_35.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_47.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_4.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_3.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_30.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_36.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_6.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_33.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_43.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_44.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_45.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_5.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_37.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_46.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_48.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_32.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_12.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_13.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_14.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_10.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_15.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_16.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_25.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_26.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_11.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_2.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_20.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_21.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_23.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_19.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_17.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_18.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_weaned/pair_24.0_weaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "for pair in pair_numbers:\n",
    "    ! qiime composition add-pseudocount \\\n",
    "    --i-table $data_dir/pairs_weaned/pair_\"$pair\"_weaned.qza \\\n",
    "    --o-composition-table $data_dir/pairs_weaned/pair_\"$pair\"_weaned_pseudo.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ed3a312e-df92-49f8-a837-30aca25da77d",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_42.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_27.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_28.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_39.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_8.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-yn1ve37p.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_40.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-qundbbo_.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-19o_hdlm.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-49h5olvy.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_3.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_30.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_36.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_6.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_33.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_43.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_44.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_45.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_5.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-lx2rnh64.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  All values in `grouping` are unique. This method cannot operate on a grouping vector with only unique values (e.g., there are no 'within' variance because each group of samples contains only a single sample).\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-1pqrurts.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-rw1pvq3u.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_32.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_12.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_13.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_14.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-wzt0irro.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_15.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_16.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_25.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-q2rig8io.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-o74kjtru.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_2.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_20.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_21.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_23.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-wvd1m1ab.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_17.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_18.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_weaned/ancom_24.0_weaned.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "for pair in pair_numbers:\n",
    "    ! qiime composition ancom \\\n",
    "    --i-table $data_dir/pairs_weaned/pair_\"$pair\"_weaned_pseudo.qza \\\n",
    "    --m-metadata-file $data_dir/meta_wean2.tsv \\\n",
    "    --m-metadata-column host_id \\\n",
    "    --p-transform-function log \\\n",
    "    --o-visualization $data_dir/ancoms_weaned/ancom_\"$pair\"_weaned.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61fbd56c-2b7b-4dce-9624-8d1cdf712092",
   "metadata": {},
   "source": [
    "#### For Not Weaned:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b67c7c7f-fdfc-4dd7-a945-4d5e9175df18",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_42.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_27.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_28.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_39.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_8.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_29.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_40.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_35.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_47.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_4.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_3.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_30.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_36.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_6.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_33.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_43.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_44.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_45.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_5.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_37.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_46.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_48.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_32.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_12.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_13.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_14.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_10.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_15.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_16.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_25.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_26.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_11.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_2.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_20.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_21.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_23.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_19.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_17.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_18.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved FeatureTable[Composition] to: data/pairs_notweaned/pair_24.0_notweaned_pseudo.qza\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "for pair in pair_numbers:\n",
    "    ! qiime composition add-pseudocount \\\n",
    "    --i-table $data_dir/pairs_notweaned/pair_\"$pair\"_notweaned.qza \\\n",
    "    --o-composition-table $data_dir/pairs_notweaned/pair_\"$pair\"_notweaned_pseudo.qza"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b85fc3b3-7d18-4fd1-98aa-c14af09c0117",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_42.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_27.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  All values in `grouping` are unique. This method cannot operate on a grouping vector with only unique values (e.g., there are no 'within' variance because each group of samples contains only a single sample).\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-6t_8myfz.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  All values in `grouping` are unique. This method cannot operate on a grouping vector with only unique values (e.g., there are no 'within' variance because each group of samples contains only a single sample).\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-t1bpvox3.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_8.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-j120i6s5.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_40.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-n7fi180g.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-pnmlf0w_.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-8ub7poc1.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_3.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_30.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_36.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_6.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_33.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_43.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_44.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_45.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_5.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-5yvgog26.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_46.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-6ecoluko.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_32.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_12.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_13.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_14.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-edqao46i.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  All values in `grouping` are unique. This method cannot operate on a grouping vector with only unique values (e.g., there are no 'within' variance because each group of samples contains only a single sample).\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-bbcqo36g.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_16.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_25.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-xam2osq4.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-f5ajzcgg.log\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  All values in `grouping` are unique. This method cannot operate on a grouping vector with only unique values (e.g., there are no 'within' variance because each group of samples contains only a single sample).\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-bffymepk.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_20.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  All values in `grouping` are unique. This method cannot operate on a grouping vector with only unique values (e.g., there are no 'within' variance because each group of samples contains only a single sample).\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-mogp1_a_.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_23.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[31m\u001b[1mPlugin error from composition:\n",
      "\n",
      "  `ids_to_keep` must contain at least one ID.\n",
      "\n",
      "Debug info has been saved to /tmp/qiime2-q2cli-err-x9r1qzk8.log\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_17.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_18.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m\u001b[32mSaved Visualization to: data/ancoms_notweaned/ancom_24.0_notweaned.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "for pair in pair_numbers:\n",
    "    ! qiime composition ancom \\\n",
    "    --i-table $data_dir/pairs_notweaned/pair_\"$pair\"_notweaned_pseudo.qza \\\n",
    "    --m-metadata-file $data_dir/meta_wean2.tsv \\\n",
    "    --m-metadata-column host_id \\\n",
    "    --p-transform-function log \\\n",
    "    --o-visualization $data_dir/ancoms_notweaned/ancom_\"$pair\"_notweaned.qzv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a60ee4ea-7ffe-4145-9a43-5e6bb0eb1e04",
   "metadata": {},
   "source": [
    "#### Some twins do not have any samples for the stages 'weaned' or 'not weaning' leading to the error messages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d136ff7e-730b-4d2f-8cad-ff4403d91ab4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mExported data/ancoms_notweaned/ancom_24.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_24.0_notweaned.tsv\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!qiime tools export \\\n",
    "  --input-path $data_dir/ancoms_notweaned/ancom_24.0_notweaned.qzv \\\n",
    "  --output-path $data_dir/ancoms_notweaned/ancom_24.0_notweaned.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d86c85cf-bc55-4d79-85e1-a3d313ead714",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-17kj25ot')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 05201a39-a4c7-425f-adb7-175c79bc5133>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(f'{data_dir}/ancoms_notweaned/ancom_24.0_notweaned.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71de5b5d-8c31-48b9-a62c-40b722f5325f",
   "metadata": {},
   "source": [
    "#### Now we have the the volcano plot and corresponding table showing the W-values for all different features. In order to make further statistical analysis, we define a threshold for the W-values and count the features above. This number reflects the difference between the twins for the corresponding stage. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6bfa742-245b-4e90-b743-77533764a1f2",
   "metadata": {},
   "source": [
    "#### For this, we need to convert the ancom results to a tsv file:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "388e5073-867e-48d0-a15f-333f02c069ba",
   "metadata": {},
   "source": [
    "#### For weaned:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c424328-fb55-4f00-931b-2768c78e08e2",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mExported data/ancoms_weaned/ancom_42.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_42.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_27.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_27.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_28.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_28.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_39.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_39.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_8.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_8.0_weaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_29.0_weaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_40.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_40.0_weaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_35.0_weaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_47.0_weaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_4.0_weaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_3.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_3.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_30.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_30.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_36.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_36.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_6.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_6.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_33.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_33.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_43.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_43.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_44.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_44.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_45.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_45.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_5.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_5.0_weaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_37.0_weaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_46.0_weaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_48.0_weaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_32.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_32.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_12.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_12.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_13.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_13.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_14.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_14.0_weaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_10.0_weaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_15.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_15.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_16.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_16.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_25.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_25.0_weaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_26.0_weaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_11.0_weaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_2.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_2.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_20.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_20.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_21.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_21.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_23.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_23.0_weaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_weaned/ancom_19.0_weaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_17.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_17.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_18.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_18.0_weaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_weaned/ancom_24.0_weaned.qzv as Visualization to directory data/ancoms_weaned/ancom_24.0_weaned.tsv\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "for pair in pair_numbers:\n",
    "    !qiime tools export \\\n",
    "      --input-path $data_dir/ancoms_weaned/ancom_\"$pair\"_weaned.qzv \\\n",
    "      --output-path $data_dir/ancoms_weaned/ancom_\"$pair\"_weaned.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2685464-3693-4ef7-8679-ff61506950eb",
   "metadata": {},
   "source": [
    "#### For not weaned:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8cf86816-bc52-40cc-a7dd-c38e3d2ce87b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mExported data/ancoms_notweaned/ancom_42.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_42.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_27.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_27.0_notweaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_28.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_39.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_8.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_8.0_notweaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_29.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_40.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_40.0_notweaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_35.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_47.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_4.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_3.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_3.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_30.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_30.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_36.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_36.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_6.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_6.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_33.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_33.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_43.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_43.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_44.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_44.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_45.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_45.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_5.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_5.0_notweaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_37.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_46.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_46.0_notweaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_48.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_32.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_32.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_12.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_12.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_13.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_13.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_14.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_14.0_notweaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_10.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_15.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_16.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_16.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_25.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_25.0_notweaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_26.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_11.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_2.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_20.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_20.0_notweaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_21.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_23.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_23.0_notweaned.tsv\u001b[0m\n",
      "Usage: \u001b[94mqiime tools export\u001b[0m [OPTIONS]\n",
      "\n",
      "  Exporting extracts (and optionally transforms) data stored inside an\n",
      "  Artifact or Visualization. Note that Visualizations cannot be transformed\n",
      "  with --output-format\n",
      "\n",
      "\u001b[1mOptions\u001b[0m:\n",
      "  \u001b[94m\u001b[4m--input-path\u001b[0m ARTIFACT/VISUALIZATION\n",
      "                        Path to file that should be exported        \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m\u001b[4m--output-path\u001b[0m PATH    Path to file or directory where data should be\n",
      "                        exported to                                 \u001b[35m[required]\u001b[0m\n",
      "  \u001b[94m--output-format\u001b[0m TEXT  Format which the data should be exported as. This\n",
      "                        option cannot be used with Visualizations\n",
      "  \u001b[94m--help\u001b[0m                Show this message and exit.\n",
      "\n",
      "\u001b[33m                    There was a problem with the command:                     \u001b[0m\n",
      "\u001b[31m\u001b[1m (1/1) Invalid value for '--input-path': File\n",
      "  'data/ancoms_notweaned/ancom_19.0_notweaned.qzv' does not exist.\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_17.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_17.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_18.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_18.0_notweaned.tsv\u001b[0m\n",
      "\u001b[32mExported data/ancoms_notweaned/ancom_24.0_notweaned.qzv as Visualization to directory data/ancoms_notweaned/ancom_24.0_notweaned.tsv\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "for pair in pair_numbers:\n",
    "    !qiime tools export \\\n",
    "      --input-path $data_dir/ancoms_notweaned/ancom_\"$pair\"_notweaned.qzv \\\n",
    "      --output-path $data_dir/ancoms_notweaned/ancom_\"$pair\"_notweaned.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a19e4916-9936-44b5-8184-950fa0749359",
   "metadata": {},
   "source": [
    "#### Next, we export the ANCOM results of each tsv to pandas so we can do further calculations:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b74d5ffb-4e40-42a9-9bdc-a7fbb1ea9bc7",
   "metadata": {},
   "source": [
    "#### Generate new dictionaries with the significant ASVs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6ef9949c-aae2-4c7b-903e-44a96e0b899b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ancoms_weaned = dict()\n",
    "for pair in pair_numbers:\n",
    "    file_exists = exists(f'{data_dir}/ancoms_weaned/ancom_'+str(pair)+'_weaned.tsv')\n",
    "    if file_exists == True:\n",
    "        ancom = pd.read_csv(data_dir + '/ancoms_weaned/ancom_'+str(pair)+'_weaned.tsv/ancom.tsv', sep = '\\t')\n",
    "        ancoms_weaned[pair] = ancom[ancom['Reject null hypothesis']==True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3e13a01c-994c-4566-9b8d-e416297d9b1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ancoms_notweaned = dict()\n",
    "for pair in pair_numbers:\n",
    "    file_exists = exists(f'{data_dir}/ancoms_notweaned/ancom_'+str(pair)+'_notweaned.tsv')\n",
    "    if file_exists == True:\n",
    "        ancom = pd.read_csv(data_dir + '/ancoms_notweaned/ancom_'+str(pair)+'_notweaned.tsv/ancom.tsv', sep = '\\t')\n",
    "        ancoms_notweaned[pair] = ancom[ancom['Reject null hypothesis']==True]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ea1e4fb-8e41-4136-9660-7672e05481fc",
   "metadata": {},
   "source": [
    "#### For the next step we need to find out which numbers belong to the mono- or dizygotic twin pairs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d67a77c7-5f58-418f-a546-7396a168ae34",
   "metadata": {},
   "outputs": [],
   "source": [
    "monos = metadata['host_id'][metadata['zygosity']=='Monozygotic']\n",
    "monozygotic_ids = list(dict.fromkeys(monos.round(0)))\n",
    "dis = metadata['host_id'][metadata['zygosity']=='Dizygotic']\n",
    "dizygotic_ids = list(dict.fromkeys(dis.round(0)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25050e33-8c01-478f-86ac-88485f99aadc",
   "metadata": {},
   "source": [
    "#### Then we can count the amount of significantly differentially abundant ASVs for the weaned and not weaned stages."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e665cbdb-4d9f-4b25-8c70-8d4bf7702256",
   "metadata": {},
   "source": [
    "#### First just for mono- and dizygotic:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "52e26034-4aff-42b5-b5f5-2de602f44602",
   "metadata": {},
   "outputs": [],
   "source": [
    "mono_numbers = dict()\n",
    "for pair in monozygotic_ids:\n",
    "    existence_weaned = exists(f'{data_dir}/ancoms_weaned/ancom_'+str(pair)+'_weaned.tsv')\n",
    "    existence_notweaned = exists(f'{data_dir}/ancoms_notweaned/ancom_'+str(pair)+'_notweaned.tsv')\n",
    "    if existence_weaned == True:\n",
    "        number_weaned = len(ancoms_weaned[pair].index)\n",
    "    if existence_notweaned == True:\n",
    "        number_notweaned = len(ancoms_notweaned[pair].index)\n",
    "    mono_numbers[pair] = number_weaned + number_notweaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "4cd97615-cd0d-4e03-b9f2-da7d5d2ee0e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "di_numbers = dict()\n",
    "for pair in dizygotic_ids:\n",
    "    existence_weaned = exists(f'{data_dir}/ancoms_weaned/ancom_'+str(pair)+'_weaned.tsv')\n",
    "    existence_notweaned = exists(f'{data_dir}/ancoms_notweaned/ancom_'+str(pair)+'_notweaned.tsv')\n",
    "    if existence_weaned == True:\n",
    "        number_weaned = len(ancoms_weaned[pair].index)\n",
    "    if existence_notweaned == True:\n",
    "        number_notweaned = len(ancoms_notweaned[pair].index)\n",
    "    di_numbers[pair] = number_weaned + number_notweaned"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a7afe1d-ddd6-474f-86e2-653908de42a1",
   "metadata": {},
   "source": [
    "#### And also for each stage:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "d084814e-f0bd-454e-9d98-324241539466",
   "metadata": {},
   "outputs": [],
   "source": [
    "mono_weaned = dict()\n",
    "for pair in monozygotic_ids:\n",
    "    existence_weaned = exists(f'{data_dir}/ancoms_weaned/ancom_'+str(pair)+'_weaned.tsv')\n",
    "    if existence_weaned == True:\n",
    "        number_weaned = len(ancoms_weaned[pair].index)\n",
    "    mono_weaned[pair] = number_weaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "3f9acd5e-ead5-4fa0-8c73-86114a0a2f1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mono_notweaned = dict()\n",
    "for pair in monozygotic_ids:\n",
    "    existence_notweaned = exists(f'{data_dir}/ancoms_notweaned/ancom_'+str(pair)+'_notweaned.tsv')\n",
    "    if existence_notweaned == True:\n",
    "        number_notweaned = len(ancoms_notweaned[pair].index)\n",
    "    mono_notweaned[pair] = number_notweaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "fc23c5d6-de57-4c9d-921f-255113295a37",
   "metadata": {},
   "outputs": [],
   "source": [
    "di_weaned = dict()\n",
    "for pair in dizygotic_ids:\n",
    "    existence_weaned = exists(f'{data_dir}/ancoms_weaned/ancom_'+str(pair)+'_weaned.tsv')\n",
    "    if existence_weaned == True:\n",
    "        number_weaned = len(ancoms_weaned[pair].index)\n",
    "    di_weaned[pair] = number_weaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "209813e6-c102-450d-a3f7-b4669e756f2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "di_notweaned = dict()\n",
    "for pair in dizygotic_ids:\n",
    "    existence_notweaned = exists(f'{data_dir}/ancoms_notweaned/ancom_'+str(pair)+'_notweaned.tsv')\n",
    "    if existence_notweaned == True:\n",
    "        number_notweaned = len(ancoms_notweaned[pair].index)\n",
    "    di_notweaned[pair] = number_notweaned"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8676fd03-abb6-42c5-8f7e-ab53e3acc19c",
   "metadata": {},
   "source": [
    "#### Get numbers into dataframe:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fb8eabb-5293-4d25-8c8e-fa7919848e5e",
   "metadata": {},
   "source": [
    "#### Just mono- and dizygotic:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "a85d008f-0e28-4160-80a9-57be8d6eb0f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    13.571429\n",
       "dtype: float64"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mono_diff = pd.DataFrame.from_dict(mono_numbers, orient='index')\n",
    "mono_diff.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "5a11e16f-6981-40b2-b38b-281312699143",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    77.777778\n",
       "dtype: float64"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "di_diff = pd.DataFrame.from_dict(di_numbers, orient='index')\n",
    "di_diff.mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16ee7043-ee66-4067-9789-4ba996759368",
   "metadata": {},
   "source": [
    "#### And also depending on stage:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "eceb2180-8789-4560-b555-bd14ec4ec157",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1.619048\n",
       "dtype: float64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mono_diff_weaned = pd.DataFrame.from_dict(mono_weaned, orient='index')\n",
    "mono_diff_weaned.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "8f70903e-de62-482f-86fa-8557a8eaf978",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    11.952381\n",
       "dtype: float64"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mono_diff_notweaned = pd.DataFrame.from_dict(mono_notweaned, orient='index')\n",
    "mono_diff_notweaned.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "b6fda61c-52b3-4345-8b3a-485d885b0ed0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    19.166667\n",
       "dtype: float64"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "di_diff_weaned = pd.DataFrame.from_dict(di_weaned, orient='index')\n",
    "di_diff_weaned.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "42df9800-630a-4edd-b86d-599c7e5048b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    58.611111\n",
       "dtype: float64"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "di_diff_notweaned = pd.DataFrame.from_dict(di_notweaned, orient='index')\n",
    "di_diff_notweaned.mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da094aa8-3feb-4296-ab5d-395107464c95",
   "metadata": {},
   "source": [
    "#### Now we have the two dataframes for each zygosity. As index we have the number of the twin pair and next to it the number of ASVs for which we could reject the null hypothesis in the ANCOM (significantly differentially abundant)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4639f52c-4018-46cb-b4e1-7581632bfd57",
   "metadata": {},
   "source": [
    "# How can the means within the stages be smaller than when just comparing zygosity?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15258e4a-120b-4fe9-863b-794713c5d92c",
   "metadata": {},
   "source": [
    "### 6. Beta diversity for visuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6e978586-6649-4fb1-9dfd-a28444b6c043",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved Visualization to: data/alpha-rarefaction2.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! qiime diversity alpha-rarefaction \\\n",
    "    --i-table $data_dir/phylogeny_filtered_table.qza \\\n",
    "    --i-phylogeny $data_dir/reference-tree.qza \\\n",
    "    --p-max-depth 10000 \\\n",
    "    --m-metadata-file $or_dir/new_metadata_for_a_diversity.tsv \\\n",
    "    --o-visualization $data_dir/alpha-rarefaction2.qzv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e96046f9-0139-40bd-8822-8bd9a97e3f35",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-dqk_6fzs')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: b35ac1ff-5d92-4097-adf1-7cb5a01826d8>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(f'{data_dir}/alpha-rarefaction2.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89c99106-b979-49fd-a4b6-e9495dd04289",
   "metadata": {},
   "source": [
    "#### Rarefaction at 1500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "665ab6e9-a230-4111-8829-6ff9521e2c4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32mSaved FeatureTable[Frequency] to: data/core-metrics-results/rarefied_table.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: data/core-metrics-results/faith_pd_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: data/core-metrics-results/observed_features_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: data/core-metrics-results/shannon_vector.qza\u001b[0m\n",
      "\u001b[32mSaved SampleData[AlphaDiversity] to: data/core-metrics-results/evenness_vector.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: data/core-metrics-results/unweighted_unifrac_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: data/core-metrics-results/weighted_unifrac_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: data/core-metrics-results/jaccard_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved DistanceMatrix to: data/core-metrics-results/bray_curtis_distance_matrix.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: data/core-metrics-results/unweighted_unifrac_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: data/core-metrics-results/weighted_unifrac_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: data/core-metrics-results/jaccard_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved PCoAResults to: data/core-metrics-results/bray_curtis_pcoa_results.qza\u001b[0m\n",
      "\u001b[32mSaved Visualization to: data/core-metrics-results/unweighted_unifrac_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: data/core-metrics-results/weighted_unifrac_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: data/core-metrics-results/jaccard_emperor.qzv\u001b[0m\n",
      "\u001b[32mSaved Visualization to: data/core-metrics-results/bray_curtis_emperor.qzv\u001b[0m\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! qiime diversity core-metrics-phylogenetic \\\n",
    "  --i-table $data_dir/phylogeny_filtered_table.qza \\\n",
    "  --i-phylogeny $data_dir/reference-tree.qza \\\n",
    "  --m-metadata-file $or_dir/metadata.tsv \\\n",
    "  --p-sampling-depth 1500 \\\n",
    "  --output-dir $data_dir/core-metrics-results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0c342d90-81fa-4de0-8ec2-1e87a1cca17a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-trimmx0c')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 431fa773-66d2-42e4-ac40-f65f5139c66d>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(f'{data_dir}/core-metrics-results/unweighted_unifrac_emperor.qzv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "98ce2915-a3a5-4684-8060-c56106ab7b25",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><img onload=\"(function(div, url){\n",
       "if (typeof require !== 'undefined') {\n",
       "    var baseURL = require.toUrl('').split('/').slice(0, -2).join('/');\n",
       "} else {\n",
       "    var baseURL = JSON.parse(\n",
       "        document.getElementById('jupyter-config-data').innerHTML\n",
       "    ).baseUrl.slice(0, -1);\n",
       "}\n",
       "url = baseURL + url;\n",
       "fetch(url).then(function(res) {\n",
       "    if (res.status === 404) {\n",
       "        div.innerHTML = 'Install QIIME 2 Jupyter extension with:<br />' +\n",
       "                        '<code>jupyter serverextension enable --py qiime2' +\n",
       "                        ' --sys-prefix</code><br />then restart your server.' +\n",
       "                        '<br /><br />(Interactive output not available on ' +\n",
       "                        'static notebook viewer services like nbviewer.)';\n",
       "    } else if (res.status === 409) {\n",
       "        div.innerHTML = 'Visualization no longer in scope. Re-run this cell' +\n",
       "                        ' to see the visualization.';\n",
       "    } else if (res.ok) {\n",
       "        url = res.url;\n",
       "        div.innerHTML = '<iframe src=\\'' + url + '\\' style=\\'' +\n",
       "                        'width: 100%; height: 700px; border: 0;\\'>' +\n",
       "                        '</iframe><hr />Open in a: <a href=\\'' + url + '\\'' +\n",
       "                        ' target=\\'_blank\\'>new window</a>'\n",
       "    } else {\n",
       "        div.innerHTML = 'Something has gone wrong. Check notebook server for' +\n",
       "                        ' errors.';\n",
       "    }\n",
       "});\n",
       "})(this.parentElement, '/qiime2/redirect?location=/tmp/qiime2-archive-it0auxy2')\" src=\"data:image/gif;base64,R0lGODlhAQABAIAAAP///wAAACH5BAEAAAAALAAAAAABAAEAAAICRAEAOw==\" /></div>"
      ],
      "text/plain": [
       "<visualization: Visualization uuid: 67705828-583e-45ac-84b4-81304be6e651>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Visualization.load(f'{data_dir}/core-metrics-results/weighted_unifrac_emperor.qzv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2d37c59-f0ac-45bd-8959-89016ef917b4",
   "metadata": {},
   "source": [
    "### 7. ANCOM for zygosity column --> find significance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdaf4b94-f059-4e31-85db-d2211ea729eb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
